up-streamer/src/route.rs

```diff
@@ -0,0 +1,106 @@
+/********************************************************************************
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use async_std::sync::Arc;
+use log::*;
+use up_rust::{UAuthority, UTransport};
+
+const ROUTE_TAG: &str = "Route:";
+const ROUTEFN_NEW_TAG: &str = "new():";
+
+///
+/// [`Route`] is defined as a combination of [`UAuthority`][up_rust::UAuthority] and
+/// [`Arc<Mutex<Box<dyn UTransport>>>`][up_rust::UTransport] as routes are at the [`UAuthority`][up_rust::UAuthority] level.
+///
+/// # Examples
+///
+/// ```
+/// use std::sync::Arc;
+/// use async_std::sync::Mutex;
+/// use up_rust::{Number, UAuthority, UTransport};
+/// use up_streamer::Route;
+///
+/// # pub mod up_client_foo {
+/// #     use std::sync::Arc;
+/// #     use up_rust::{UMessage, UTransport, UStatus, UUIDBuilder, UUri, UListener};
+/// #     use async_trait::async_trait;
+/// #     pub struct UPClientFoo;
+/// #
+/// #     #[async_trait]
+/// #     impl UTransport for UPClientFoo {
+/// #         async fn send(&self, _message: UMessage) -> Result<(), UStatus> {
+/// #             todo!()
+/// #         }
+/// #
+/// #         async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+/// #             todo!()
+/// #         }
+/// #
+/// #         async fn register_listener(
+/// #             &self,
+/// #             topic: UUri,
+/// #             _listener: Arc<dyn UListener>,
+/// #         ) -> Result<(), UStatus> {
+/// #             println!("UPClientFoo: registering topic: {:?}", topic);
+/// #             Ok(())
+/// #         }
+/// #
+/// #         async fn unregister_listener(&self, topic: UUri, _listener: Arc<dyn UListener>) -> Result<(), UStatus> {
+/// #             println!(
+/// #                 "UPClientFoo: unregistering topic: {topic:?}"
+/// #             );
+/// #             Ok(())
+/// #         }
+/// #     }
+/// #
+/// #     impl UPClientFoo {
+/// #         pub fn new() -> Self {
+/// #             Self {}
+/// #         }
+/// #     }
+/// # }
+///
+/// let local_transport: Arc<Box<dyn UTransport>> = Arc::new(Box::new(up_client_foo::UPClientFoo::new()));
+///
+/// let authority_foo = UAuthority {
+///     name: Some("foo_name".to_string()).into(),
+///     number: Some(Number::Ip(vec![192, 168, 1, 100])),
+///     ..Default::default()
+/// };
+///
+/// let local_route = Route::new("local_route", authority_foo, local_transport);
+/// ```
+#[derive(Clone)]
+pub struct Route {
```

Comment:
IIRC we had already discussed the naming in the review meeting a few weeks ago.

IMHO the term _Route_ in this context is misleading because the structure only represents one end of a route. I do not have a good proposal but it seems to be closer to an Endpoint of some kind than a Route ...

Reply:
Yeah I think we ended with "I will await further discussion on the PR for better ideas" :hugs: 

I like your idea of calling what's currently a `Route` an `Endpoint`.

Open to other suggestions as well.

Reply:
Went ahead and updated to `Endpoint` for now :+1: